### PR TITLE
Add corr and cov for dask.dataframe

### PR DESCRIPTION
**Pull request by jcrist**
Implements `corr` and `cov` methods for both `DataFrame` and `Series` objects. Note that only Pearson correlation is implemented, the other methods require a sort. The computation is done in a single pass, and is numerically stable.

Partially addresses #1000.

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -2152,6 +2181,99 @@ def quantile(df, q):
...
+        m = np.nansum((x - sums/np.where(counts, counts, np.nan))**2, 0)
+        dtype.append(('m', m.dtype))
+
+    out = np.empty(counts.shape, dtype=dtype)
+    out['sum'] = sums
+    out['count'] = counts
+    out['cov'] = cov * (counts - 1)
+    if corr:
+        out['m'] = m
+    return out
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Perhaps return a tuple rather than an array of structs?

Comment (mrocklin) [inline dask/dataframe/core.py]: How large is this intermediate result?  It seems like it might be close to the size of the input?

Comment (jcrist) [inline dask/dataframe/core.py]: I thought about doing that, and could go either way. This made it a bit easier to work with both types of intermediates, and serializes slightly faster. The decision between the two was mostly arbitrary though.

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -2152,6 +2181,99 @@ def quantile(df, q):
...
+    totals = np.cumsum(sums, 0)
+    ns = np.cumsum(counts, 0)
+
+    ts = ns[:-1] * sums[1:] - counts[1:] * totals[:-1]
+    dens = ns[1:] * ns[:-1] * counts[1:]
+    dens = np.where(dens, dens, np.nan)
+    C = (np.nansum(data['cov'], 0) +
+         np.nansum(ts * ts.transpose((0, 2, 1)) / dens, 0))
+    C[ns[-1] < min_periods] = np.nan
+    nobs = np.where(ns[-1], ns[-1], np.nan)
```
Comment (mrocklin) [inline dask/dataframe/core.py]: If the intermediate results are sizable, then how much of a pain would it be to share the small boundary elements and do some of the aggregation on the chunk side?

File: dask/dataframe/core.py (hunk 1 above)
Comment (jcrist) [inline dask/dataframe/core.py]: Intermediate result shape is `(len(df.columns), len(df.columns))`.

File: dask/dataframe/tests/test_dataframe.py [hunk 3]
```diff
@@ -1410,6 +1410,49 @@ def test_apply():
...
+    assert ddf.cov(10)._name != ddf.cov()._name
+
+    a = df.A
+    b = df.B
+    da = dd.from_pandas(a, npartitions=3)
+    db = dd.from_pandas(b, npartitions=4)
+    assert eq(da.cov(db), a.cov(b))
+    assert eq(da.cov(db, 10), a.cov(b, 10))
+    assert da.cov(db)._name == da.cov(db)._name
+    assert da.cov(db, 10)._name != da.cov(db)._name
```
Comment (mrocklin) [inline dask/dataframe/tests/test_dataframe.py]: Nice reuse of `concat`.  This is slick.

File: dask/dataframe/tests/test_dataframe.py [hunk 4]
```diff
@@ -1410,6 +1410,49 @@ def test_apply():
...
+def test_corr():
+    df = pd.util.testing.makeMissingDataframe(0.3, 42)
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    assert eq(ddf.corr(), df.corr())
+    assert eq(ddf.corr(min_periods=10), df.corr(min_periods=10))
+    assert ddf.corr()._name == ddf.corr()._name
+    assert ddf.corr(min_periods=10)._name != ddf.corr()._name
+
+    pytest.raises(NotImplementedError, lambda: ddf.corr(method='spearman'))
```
Comment (mrocklin) [inline dask/dataframe/tests/test_dataframe.py]: Apparently there si a context manager version of this if you're interested.

```python
with pytest.raises(NotImplementedError):
    ddf.corr(method='spearman')
```

**Comment by mrocklin** [comment]
I have to admit a certain amount of ignorance here.  I'm not familiar with the algorithm.

File: dask/dataframe/core.py (hunk 1 above)
Comment (mrocklin) [inline dask/dataframe/core.py]: Ah!  Well then please ignore most of my comments :)

File: dask/dataframe/core.py (hunk 2 above)
Comment (jcrist) [inline dask/dataframe/core.py]: We're being a bit wasteful here as the intermediates are symmetric - it's much easier to ignore this until it becomes a problem though. For a 100 column dataframe the intermediates are 320 KB. A 10 column dataframe is only 3.2 KB. I'd probably switch to using a tree reduction before I messed with the symmetry. Besides this, no more reduction can be done inside the chunk.

Comment (mrocklin) [inline dask/dataframe/core.py]: Yeah, please ignore this comment.  I was under the incorrect assumption that the intermediates were sizable.  I agree that optimization at this point, especially in the suggested direction, would be premature.

**Pull request merged**